ddtrace/propagation/http.py

```diff
@@ -15,16 +23,279 @@
 HTTP_HEADER_PARENT_ID = "x-datadog-parent-id"
 HTTP_HEADER_SAMPLING_PRIORITY = "x-datadog-sampling-priority"
 HTTP_HEADER_ORIGIN = "x-datadog-origin"
+HTTP_HEADER_B3_SINGLE = "b3"
+HTTP_HEADER_B3_TRACE_ID = "x-b3-traceid"
+HTTP_HEADER_B3_SPAN_ID = "x-b3-spanid"
+HTTP_HEADER_B3_SAMPLED = "x-b3-sampled"
+HTTP_HEADER_B3_FLAGS = "x-b3-flags"
+
+
+def _possible_header(header):
+    # type: (str) -> frozenset[str]
+    return frozenset([header, get_wsgi_header(header).lower()])
 
 
 # Note that due to WSGI spec we have to also check for uppercased and prefixed
 # versions of these headers
-POSSIBLE_HTTP_HEADER_TRACE_IDS = frozenset([HTTP_HEADER_TRACE_ID, get_wsgi_header(HTTP_HEADER_TRACE_ID).lower()])
-POSSIBLE_HTTP_HEADER_PARENT_IDS = frozenset([HTTP_HEADER_PARENT_ID, get_wsgi_header(HTTP_HEADER_PARENT_ID).lower()])
-POSSIBLE_HTTP_HEADER_SAMPLING_PRIORITIES = frozenset(
-    [HTTP_HEADER_SAMPLING_PRIORITY, get_wsgi_header(HTTP_HEADER_SAMPLING_PRIORITY).lower()]
-)
-POSSIBLE_HTTP_HEADER_ORIGIN = frozenset([HTTP_HEADER_ORIGIN, get_wsgi_header(HTTP_HEADER_ORIGIN).lower()])
+POSSIBLE_HTTP_HEADER_TRACE_IDS = _possible_header(HTTP_HEADER_TRACE_ID)
+POSSIBLE_HTTP_HEADER_PARENT_IDS = _possible_header(HTTP_HEADER_PARENT_ID)
+POSSIBLE_HTTP_HEADER_SAMPLING_PRIORITIES = _possible_header(HTTP_HEADER_SAMPLING_PRIORITY)
+POSSIBLE_HTTP_HEADER_ORIGIN = _possible_header(HTTP_HEADER_ORIGIN)
+POSSIBLE_HTTP_HEADER_B3_SINGLE_HEADER = _possible_header(HTTP_HEADER_B3_SINGLE)
+POSSIBLE_HTTP_HEADER_B3_TRACE_IDS = _possible_header(HTTP_HEADER_B3_TRACE_ID)
+POSSIBLE_HTTP_HEADER_B3_SPAN_IDS = _possible_header(HTTP_HEADER_B3_SPAN_ID)
+POSSIBLE_HTTP_HEADER_B3_SAMPLEDS = _possible_header(HTTP_HEADER_B3_SAMPLED)
+POSSIBLE_HTTP_HEADER_B3_FLAGS = _possible_header(HTTP_HEADER_B3_FLAGS)
+
+
+def _extract_header_value(possible_header_names, headers, default=None):
+    # type: (FrozenSet[str], Dict[str, str], Optional[str]) -> Optional[str]
+    for header in possible_header_names:
+        try:
+            return headers[header]
+        except KeyError:
+            pass
+
+    return default
+
+
+def _normalize_headers(headers):
+    # type: (Dict[str, str]) -> Dict[str, str]
+    return {name.lower(): v for name, v in headers.items()}
+
+
+def _b3_id_to_dd_id(b3_id):
+    # type: (str) -> int
+    """Helper to convert B3 trace/span hex ids into Datadog compatible ints
+
+    If the id is > 64 bit then truncate the trailing 64 bit.
+
+    "463ac35c9f6413ad48485a3953bb6124" -> "48485a3953bb6124" -> 5208512171318403364
+    """
+    if len(b3_id) > 16:
+        return int(b3_id[-16:], 16)
+    return int(b3_id, 16)
```

Comment:
```suggestion
    return int(b3_id[-16:], 16)
```
This seems to be slightly faster on the `> 16` branch (likely due to avoiding the `if` check)
~~~
286 ns ± 6.6 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
243 ns ± 2.71 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
~~~
No difference for the `<= 16` branch:
~~~
196 ns ± 0.862 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
195 ns ± 1.6 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
~~~